### PR TITLE
[1.16.x] Environment tick event

### PR DESCRIPTION
**Pull request by mrp-v2**
This event provides a way for mods to perform weather and random tick based actions at the same time as vanilla (not actually the same time, either just before or just after).
Test mod included.

### CODE REVIEW

**Comment by KnightMiner** [comment]
Why not make the pre event cancelable, canceling can be used to prevent the random tick from being processed in vanilla logic.

Also, can you give an example usage of this event?

**Comment by mrp-v2** [comment]
Canceling the event would also cancel lightning, skeleton horse traps, water freezing in cold biomes, snow layers generating during snow, blocks like cauldrons filling with water during rain, as well as random ticks. Also, other tick events aren't cancelable, so it would be inconsistent for this one to be.

An example usage would be if a mod that adds custom biomes/world generation and wanted to have a custom precipitation that caused a block to form on the surface of the world, like snow causes snow to form. Or if a mod wanted to give blocks extra random ticks. If a mod wants something to be caused by weather, this event would be the way to do it.

The specific inspiration for this event is that I have a mod in which I want to modify snowy weather so that snow layers can accumulate multiple layers instead of being limited to just 1 layer.

File: patches/minecraft/net/minecraft/world/server/ServerChunkProvider.java.patch [hunk 1]
```diff
@@ -0,0 +1,13 @@
+--- a/net/minecraft/world/server/ServerChunkProvider.java
++++ b/net/minecraft/world/server/ServerChunkProvider.java
+@@ -358,8 +358,9 @@
+                      if (flag1 && (this.field_217246_l || this.field_217247_m) && this.field_73251_h.func_175723_af().func_177730_a(chunk.func_76632_l())) {
+                         WorldEntitySpawner.func_234979_a_(this.field_73251_h, chunk, worldentityspawner$entitydensitymanager, this.field_217247_m, this.field_217246_l, flag2);
+                      }
+-
```
Comment (ichttt) [inline patches/minecraft/net/minecraft/world/server/ServerChunkProvider.java.patch]: Please keep the newline to reduce patch size

File: src/test/java/net/minecraftforge/debug/world/EnvironmentTickEventTest.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import net.minecraftforge.fml.common.Mod;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+@Mod("environmenttickeventtest")
+public class EnvironmentTickEventTest
+{
+    private final Logger LOGGER;
+    
+    public EnvironmentTickEventTest(){
```
Comment (ichttt) [inline src/test/java/net/minecraftforge/debug/world/EnvironmentTickEventTest.java]: Please put the brackets on a new line

File: src/test/java/net/minecraftforge/debug/world/EnvironmentTickEventTest.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+        LOGGER = LogManager.getLogger();
+        MinecraftForge.EVENT_BUS.register(this);
+    }
+    
+    @SubscribeEvent
+    public void environmentTickEvent(final TickEvent.WorldTickEvent.EnvironmentTickEvent event){
```
Comment (ichttt) [inline src/test/java/net/minecraftforge/debug/world/EnvironmentTickEventTest.java]: Same here. please fix the code style

**Review by ichttt — COMMENTED** [review]
Some code style notes

**Comment by diesieben07** [comment]
Extending WorldTickEvent is a breaking change. Mods subscribed to WorldTickEvent will now receive this event as well, which will throw off tick-based timing.

**Comment by mrp-v2** [comment]
Should I extend TickEvent instead?

File: src/main/java/net/minecraftforge/fml/hooks/BasicEventHooks.java [hunk 4]
```diff
@@ -119,4 +120,14 @@ public static void onPostServerTick()
...
     }
+
+    public static void onPreEnvironmentTick(World world, Chunk chunk, int randomTickSpeed)
+    {
+        MinecraftForge.EVENT_BUS.post(new TickEvent.WorldTickEvent.EnvironmentTickEvent(LogicalSide.SERVER, TickEvent.Phase.START, world, chunk, randomTickSpeed));
+    }
+    
+    public static void onPostEnvironmentTick(World world, Chunk chunk, int randomTickSpeed)
+    {
+        MinecraftForge.EVENT_BUS.post(new TickEvent.WorldTickEvent.EnvironmentTickEvent(LogicalSide.SERVER, TickEvent.Phase.END, world, chunk, randomTickSpeed));
```
Comment (alcatrazEscapee) [inline src/main/java/net/minecraftforge/fml/hooks/BasicEventHooks.java]:
```java
new TickEvent.WorldTickEvent.EnvironmentTickEvent
```
looks wrong here, seeing that `EnvironmentTickEvent` now directly extends `TickEvent`. It should just be `new TickEvent.EnvironmentTickEvent()` for simplicity, even if this does technically compile just fine.

**Comment by sciwhiz12** [comment]
Hello! @mrp-v2, please leave a comment whether you wish to continue this PR for 1.16. Note that there is already an updated version of the feature intended by this PR at #8314, so it is your decision whether to discontinue updating this PR in favor of that or to continue with your own implementation. (We suggest the former so there is a concentration of effort in one PR rather than fragmented across multiple.) If we do not hear back from you after some time, this PR will be closed for inactivity.
If anyone else wishes to continue the work of this PR, please create a PR with a link to this PR for reference. 
Thank you for your contributions. 👋

**Comment by sciwhiz12** [comment]
Hello! Since the application of the Stale label, there has been no recent activity on this PR. As said in our previous comment, this PR will now be closed for inactivity.

If you wish to continue this PR, please open a new PR against the Latest supported version (to be conformant with our version support and contributions policy). 

As a note, PR #8314 seems to implement a similar feature to this PR.

Thank you for your contributions. 👋

**Pull request closed without merging**